api/METRICS-DASHBOARD.md

```diff
@@ -0,0 +1,73 @@
+# Proposal for metrics/dashboard API
+
+Goals:
+* Unified treatment of many different kinds of metrics in dashboard
+* A clean separation between dashboard (for visualization) and different kinds of backends (for metric calculation)
+
+Status quo:
+* Dashboard currently keeps track of the metric values in the cache dictionary `fairlearn.widget._fairlearn_widget.FairlearnWidget._response`
+* The dictionary is updated in `fairlearn.widget._fairlearn_dashboard.FairlearnDashboard._on_request`
+* There's a PR out to fill the whole data structure in  `fairlearn.metrics.create_dashboard_dictionary`
+
+Issues with the status quo:
+* Code duplication / redundancy / brittleness due to copy-paste errors
+* Currently only one kind of a metric (`<metric>_summary`) is supported and hard-wired into the dashboard dictionary
+
+## Proposal
+
+### Part I: More general dashboard dictionary
+
+```
+{
+  "prediction_type":  "binary_classification" or "probabilistic_binary_classification" or "regression",
+  "y_true": float[],
+  "y_pred": float[][],
+  "sensitive_features": [
+    {
+      "column_name": String,

```

Comment:
It's a far bigger change, but my personal preference would be to have dictionaries instead of arrays. I feel that the nested `precomputedMetrics` dictionary is a bit fragile right now. For example, my round trip for the AzureML integration only works if the models and sensitive features are sorted by name, since when the 2D array is being reassembled, I have to pick _some_ ordering of the models and sensitive features. To make things deterministic, I sort, but that means that while the reconstructed dashboard is logically the same, it's not physically the same.

Reply:
Actually, maybe not. I don't think there really is a good way of doing this in JSON, and the list of lists works as well as nesting dictionaries.